File: src/main/java/no/ntnuf/towlog/towlog2/main/MainActivity.java
package no.ntnuf.towlog.towlog2.main;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AutoCompleteTextView;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.OptionalDataException;
import java.io.StreamCorruptedException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Locale;

import no.ntnuf.towlog.towlog2.common.ColoringUtil;
import no.ntnuf.towlog.towlog2.common.Contact;
import no.ntnuf.towlog.towlog2.dayoverview.DayOverviewActivity;
import no.ntnuf.tow.towlog2.R;
import no.ntnuf.towlog.towlog2.common.ContactListManager;
import no.ntnuf.towlog.towlog2.common.DayLog;
import no.ntnuf.towlog.towlog2.fiken.FikenContactRequestTask;

/**
 * An example full-screen activity that shows and hides the system UI (i.e.
 * status bar and navigation/system bar) with user interaction.
 */
public class MainActivity extends AppCompatActivity {

    private final String dayLogFileNamePrefix = "daylog_";
    private String dayLogFileNameSuffix = "";


    private Button startDayButton;
    private Button resumeDayButton;
    private DatePicker datepicker;

    private Toolbar toolbar;
    private AlertDialog daylogdialog;
    private AlertDialog loadfikencontactsdialog;

    private Contact selectedTowPilot;
    private Date selecteddate = new Date();

    private ArrayList<String> countries = new ArrayList<String>();

    private ContactListManager contactlistmanager;
    private AutoCompleteTextView towPilotNameIn;
    private ImageView towPilotCheckmark;
    private EditText towPlaneIn;

    private boolean found_daylog = false;
    private DayLog daylog;

    private SharedPreferences settings;

    private boolean autoLoadLog = true;

    private Calendar today;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        toolbar = (Toolbar) findViewById(R.id.toolbarmain);
        toolbar.setTitle("Prepare Towing");
        setSupportActionBar(toolbar);

        settings = PreferenceManager.getDefaultSharedPreferences(this);

        // Set up the tow pilot name input, including autocomplete and hasAccount check
        contactlistmanager = new ContactListManager(this);

        towPilotCheckmark = (ImageView) findViewById(R.id.towPilotCheckmark);

        towPilotNameIn = (AutoCompleteTextView) findViewById(R.id.towPilotNameIn);
        towPilotNameIn.setAdapter(contactlistmanager.getContactNameListAdapter());
        towPilotNameIn.addTextChangedListener(new TextWatcher() {
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                Contact selected = contactlistmanager.findContactFromName(String.valueOf(s));
                selectedTowPilot = selected;
                if (selected == null) {
                    towPilotCheckmark.setBackgroundResource(0);
                } else {
                    if (selected.hasAccount) {
                        towPilotCheckmark.setBackgroundResource(R.mipmap.green_checkmark);
                    } else {
                        towPilotCheckmark.setBackgroundResource(R.mipmap.new_icon_blue);
                    }
                }
            }

            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            public void afterTextChanged(Editable s) {
            }
        });


        // Set up towplane input
        towPlaneIn = (EditText) findViewById(R.id.towPlaneIn);
        towPlaneIn.setText(settings.getString("lasttowplane", settings.getString("towplane_default_reg","")));


        // Set up Datepicker
        datepicker = (DatePicker) findViewById(R.id.datePicker);
        today = Calendar.getInstance();
        datepicker.init(today.get(Calendar.YEAR), today.get(Calendar.MONTH), today.get(Calendar.DAY_OF_MONTH), new DatePicker.OnDateChangedListener() {
            @Override
            public void onDateChanged(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                Calendar c = Calendar.getInstance();
                c.set(year, monthOfYear, dayOfMonth);
                selecteddate = c.getTime();

                dayLogFileNameSuffix = String.valueOf(year) + "_" + String.valueOf(monthOfYear + 1) +
                        "_" + String.valueOf(dayOfMonth);

                Log.e("DATEPICKER", dayLogFileNameSuffix);

                if (loadDayLog()) {
                    found_daylog = true;
                    resumeDayButton.setVisibility(View.VISIBLE);
                } else {
                    found_daylog = false;
                    resumeDayButton.setVisibility(View.INVISIBLE);
                }
            }
        });

        Log.e("DATEPICKER", "Inited object, " + datepicker);

        // Start new day button
        startDayButton = (Button) findViewById(R.id.startDayButton);
        ColoringUtil.colorMe(startDayButton, getResources().getColor(R.color.colorPrimary));
        startDayButton.setTextColor(getResources().getColor(R.color.white));
        startDayButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                if (found_daylog) {
                    new AlertDialog.Builder(MainActivity.this)
                            .setMessage("There is already a log for this date. " +
                                    "Are you sure you want to overwrite it?")
                            .setCancelable(false)
                            .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog, int id) {
                                    startNewDay();
                                }
                            })
                            .setNegativeButton("No", null)
                            .show();

                } else {
                    startNewDay();
                }
            }
        });


        // Resume day button
        resumeDayButton = (Button) findViewById(R.id.resumeDayButton);
        ColoringUtil.colorMe(resumeDayButton, getResources().getColor(R.color.resumeday_button));
        resumeDayButton.setTextColor(getResources().getColor(R.color.white));
        resumeDayButton.setVisibility(View.INVISIBLE);
        resumeDayButton.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                String action = "resume";
                Intent intent = new Intent(MainActivity.this, DayOverviewActivity.class);
                Bundle bundle = bundleDayInfo();
                bundle.putSerializable("action", action);
                bundle.putSerializable("reason", "Resuming day");
                intent.putExtras(bundle);
                startActivityForResult(intent, 0);
            }
        });



        // Various dialogs for the menus
        daylogdialog = getPrevLogsAlertDialog();
        loadfikencontactsdialog = getLoadFikenContactsAlertDialog();

    }

    // Helper function to start a new day
    private void startNewDay() {
        // Add the pilot so we can autocomplete it later
        contactlistmanager.saveContact(towPilotNameIn.getText().toString().trim());

        // Add tow info to the bundle
        String action = "new";
        Intent intent = new Intent(MainActivity.this, DayOverviewActivity.class);
        Bundle bundle = bundleDayInfo();
        bundle.putSerializable("action", action);

        // Start next activity, DayOverview
        intent.putExtras(bundle);
        startActivity(intent);
    }

    // Helper function to oncreate, get an alertdialog for showing available logs
    private AlertDialog getPrevLogsAlertDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Previous logs");
        final CharSequence[] daysavailable = availableDayLogs();
        builder.setItems(daysavailable, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dayLogFileNameSuffix = getFilenameFromListpos(which);

                // Update the date fields passed on to dayoverviewactivity
                String[] daylogfilenamesplit = dayLogFileNameSuffix.split("_");
                int year = Integer.valueOf(daylogfilenamesplit[0]);
                int monthOfYear = Integer.valueOf(daylogfilenamesplit[1]) - 1;
                int dayOfMonth = Integer.valueOf(daylogfilenamesplit[2]);

                Calendar c = Calendar.getInstance();
                c.set(year, monthOfYear, dayOfMonth);
                selecteddate = c.getTime();

                // Call the resume day button which handles the dayoverview activity starting
                resumeDayButton.callOnClick();
            }
        });
        return builder.create();
    }

    // Helper function to oncreate, get an alertdialog for loading contacts from fiken
    private AlertDialog getLoadFikenContactsAlertDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Load Contacts from Fiken");
        builder.setMessage("Connecting to Fiken...");
        return builder.create();
    }

    // Helper function
    // Get the filename suffix from an index, corresponding to the position
    // of the file list
    private String getFilenameFromListpos(int which) {
        String[] listOfFiles = this.getFilesDir().list();
        List<String> arrayOfFiles = new ArrayList<String>(Arrays.asList(listOfFiles));
        Collections.sort(arrayOfFiles);

        int c = 0;
        for (int i = 0; i < arrayOfFiles.size(); i++) {
            String name = arrayOfFiles.get(i);
            // Count only the files that contain daylogs
            if (name.startsWith(dayLogFileNamePrefix)) {
                if (c == which) {
                    return name.replaceFirst(dayLogFileNamePrefix, "");
                }
                c++;
            }
        }
        return "";
    }

    // Search files for available logs
    private CharSequence[] availableDayLogs() {
        String[] listOfFiles = this.getFilesDir().list();
        ArrayList<String> arrayOfFiles = new ArrayList<String>(Arrays.asList(listOfFiles));
        Collections.sort(arrayOfFiles);

        ArrayList<String> s = new ArrayList<String>();

        SimpleDateFormat df = new SimpleDateFormat("yyyy M d", Locale.ENGLISH);
        SimpleDateFormat outdf = new SimpleDateFormat("cccc d/M/yyyy");

        for (int i = 0; i < arrayOfFiles.size(); i++) {
            String name = arrayOfFiles.get(i);
            // Add only the files that contain daylogs
            if (name.startsWith(dayLogFileNamePrefix)) {
                // Prettify it
                String[] namesplit = name.split("_");
                if (namesplit.length == 4) {

                    Date result = null;
                    try {
                        String parsestring = namesplit[1] + " " + namesplit[2] + " " + namesplit[3];
                        result = df.parse(parsestring);
                        String print = outdf.format(result);
                        s.add(print);
                    } catch (ParseException e) {
                        s.add("Parseerror, name: "+name);
                    }

                } else {
                    // Messed up formatting of the filename
                    s.add("Parse error, split: "+name);
                }
            }
        }

        final CharSequence[] cs = s.toArray(new String[s.size()]);
        return cs;
    }

    // Helper function to pack up the tow pilot info and the date
    private Bundle bundleDayInfo() {
        Bundle bundle = new Bundle();
        Contact towpilot;
        if (selectedTowPilot == null) {
            towpilot = new Contact();
            towpilot.name = towPilotNameIn.getText().toString().trim();
        } else {
            towpilot = selectedTowPilot;
        }
        String towplane = towPlaneIn.getText().toString().trim();
        Date d = selecteddate;

        bundle.putSerializable("towpilot", towpilot);
        bundle.putSerializable("towplane", towplane);
        bundle.putSerializable("date", d);
        return bundle;
    }

    // Helper function to load the day log from a file
    private boolean loadDayLog() {
        boolean res = false;
        String fullfilename = dayLogFileNamePrefix + dayLogFileNameSuffix;
        try {
            FileInputStream fis = this.openFileInput(fullfilename);
            ObjectInputStream is = new ObjectInputStream(fis);
            daylog = (DayLog) is.readObject();
            is.close();
            fis.close();
            res = true;
            //Log.e("Main Load", "Existing log for this date found: " + fullfilename);
        } catch (ClassNotFoundException e) {
            e.printStackTrace();
            this.deleteFile(dayLogFileNamePrefix + dayLogFileNameSuffix);
            Toast.makeText(MainActivity.this, "Broken file selected, deleting it...",
                    Toast.LENGTH_LONG).show();
        } catch (OptionalDataException e) {
            e.printStackTrace();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
            Log.e("MAIN", "Logload, No log for this date found " + fullfilename);
        } catch (StreamCorruptedException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return res;
    }

    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_menu, menu);
        return true;
    }

    // Menu selector. Starts async tasks if needed.
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            // Show the list of days
            case R.id.menu_listtowlogs:
                daylogdialog.show();
                return true;

            case R.id.menu_loadfikencontacts:
                FikenContactRequestTask fikenRequest = new FikenContactRequestTask();
                fikenRequest.setContext(this);
                fikenRequest.setDialog(loadfikencontactsdialog);
                fikenRequest.setContactListManager(contactlistmanager);
                fikenRequest.execute();
                return true;

            case R.id.menu_settings:
                Intent intent = new Intent(MainActivity.this, SettingsActivity.class);
                startActivity(intent);
                return true;


            // Nothing to do
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    // The main activity behaves differently depending on how we return from it from other activities
    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        Log.e("MAIN", "onActivityResult() called");

        if (data != null) {
            Bundle bundle = data.getExtras();
            String action = (String) bundle.getSerializable("action");
            if (action.equals("backbutton")) {
                // Close app immediately if we return due to a back button press
                finish();
                return;
            } else if (action.equals("backtocalendarmenu")) {
                // If we return back from daylog, stay on this activity
                autoLoadLog = false;
                return;
            }
        }

        // If you got here otherwise, you came from the "back to calendar" menu option in daylog
        // If we return back from daylog, stay on this activity
        //autoLoadLog = false;
    }

    // Refresh the view when going back
    protected void onResume() {
        super.onResume();
        Log.e("MAIN", "onResume() called");

        today = Calendar.getInstance();

        // Day log loader, check if we already have a log for the current date
        dayLogFileNameSuffix = String.valueOf(datepicker.getYear()) + "_" +
                String.valueOf(datepicker.getMonth()+1) + "_" +
                String.valueOf(datepicker.getDayOfMonth());
        if (loadDayLog()) {
            found_daylog = true;
            resumeDayButton.setVisibility(View.VISIBLE);

            // Jump straight to daylog if we are currently on the shown date from a clean startup
            int day = datepicker.getDayOfMonth();
            int month = datepicker.getMonth();
            int year = datepicker.getYear();
            Calendar date_picked = Calendar.getInstance();
            date_picked.set(year, month, day);

            if (autoLoadLog && date_picked.get(Calendar.YEAR) == today.get(Calendar.YEAR) &&
                    date_picked.get(Calendar.DAY_OF_YEAR) == today.get(Calendar.DAY_OF_YEAR)) {
                resumeDayButton.callOnClick();
            }
        } else {
            found_daylog = false;
            resumeDayButton.setVisibility(View.INVISIBLE);
        }

        // Subsequent resumes will autoload logs again
        autoLoadLog = true;

        // Reload the contact list
        contactlistmanager = new ContactListManager(this);
        towPilotNameIn.setAdapter(contactlistmanager.getContactNameListAdapter());

        // Set the daylog dialog
        daylogdialog = getPrevLogsAlertDialog();

        // To avoid focus going to the text fields
        datepicker.requestFocus();

    }

}

File: app/src/main/java/no/ntnuf/tow/towlog2/MainActivity.kt
package no.ntnuf.tow.towlog2

import android.Manifest
import android.app.AlertDialog
import android.app.DatePickerDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import androidx.activity.enableEdgeToEdge
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.preference.PreferenceManager
import kotlinx.coroutines.launch
import no.ntnuf.tow.towlog2.model.ContactListManager
import no.ntnuf.tow.towlog2.viewmodel.TowingViewModel
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import kotlin.math.max

class MainActivity : AppCompatActivity() {
    private val dayLogFileNamePrefix = "daylog_"
    private val locationPermissions = arrayOf(
        Manifest.permission.ACCESS_FINE_LOCATION,
        Manifest.permission.ACCESS_COARSE_LOCATION
    )
    private val viewModel: TowingViewModel by viewModels()

    private lateinit var toolbar: Toolbar
    private lateinit var selectDateButton: Button
    private lateinit var towPilotInput: AutoCompleteTextView
    private lateinit var towPilotCheckmark: ImageView
    private lateinit var towPlaneInput: EditText
    private lateinit var mainButtonContainer: View
    private lateinit var startNewDayButton: Button
    private lateinit var resumeDayButton: Button
    private lateinit var contactAdapter: ArrayAdapter<String>

    private var isUpdatingTowPlaneText = false

    private val locationPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { results ->
        val granted = results.any { it.value }
        if (!granted) {
            val permanentlyDenied = locationPermissions.all { permission ->
                !ActivityCompat.shouldShowRequestPermissionRationale(this, permission)
            }
            if (permanentlyDenied) {
                showLocationPermissionPermanentlyDeniedDialog()
            } else {
                showLocationPermissionDeniedDialog()
            }
        }
    }

    private data class AvailableDayLog(
        val fileNameSuffix: String,
        val displayName: String
    )

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        val windowInsetsController = WindowCompat.getInsetsController(window, window.decorView)
        windowInsetsController.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        windowInsetsController.hide(WindowInsetsCompat.Type.systemBars())
        setContentView(R.layout.activity_main)

        toolbar = findViewById(R.id.toolbarmain)
        toolbar.title = getString(R.string.main_prepare_towing_title)
        setSupportActionBar(toolbar)

        bindViews()
        bindViewModel()
        bindInputHandlers()

        // Trigger initial day-log existence check for the default selected date.
        viewModel.updateSelectedDate(viewModel.selectedDate.value)

        ensureLocationPermissionOnLaunch()
    }

    override fun onResume() {
        super.onResume()
        val defaultTowPlane = (PreferenceManager.getDefaultSharedPreferences(this)
            .getString("towplane_default_reg", "") ?: "")
            .uppercase(Locale.ROOT)
        towPlaneInput.setText(defaultTowPlane)
        towPlaneInput.setSelection(defaultTowPlane.length)
        viewModel.updateTowPlane(defaultTowPlane)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.menu_previous_logs -> {
                showPreviousLogsDialog()
                return true
            }

            R.id.menu_settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun bindViews() {
        selectDateButton = findViewById(R.id.selectDateButton)
        towPilotInput = findViewById(R.id.towPilotNameIn)
        towPilotCheckmark = findViewById(R.id.towPilotCheckmark)
        towPlaneInput = findViewById(R.id.towPlaneIn)
        mainButtonContainer = findViewById(R.id.mainButtonContainer)
        startNewDayButton = findViewById(R.id.startDayButton)
        resumeDayButton = findViewById(R.id.resumeDayButton)
        applyImeAwareMarginsToBottomEnd(mainButtonContainer)

        contactAdapter = ArrayAdapter(this, android.R.layout.simple_dropdown_item_1line, mutableListOf())
        towPilotInput.setAdapter(contactAdapter)
    }

    private fun applyImeAwareMarginsToBottomEnd(view: View) {
        val marginLayoutParams = view.layoutParams as? ViewGroup.MarginLayoutParams ?: return
        val initialBottom = marginLayoutParams.bottomMargin
        val initialEnd = marginLayoutParams.marginEnd
        val initialRight = marginLayoutParams.rightMargin

        ViewCompat.setOnApplyWindowInsetsListener(view) { target, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            val ime = insets.getInsets(WindowInsetsCompat.Type.ime())
            val updated = target.layoutParams as? ViewGroup.MarginLayoutParams
                ?: return@setOnApplyWindowInsetsListener insets

            updated.bottomMargin = initialBottom + max(systemBars.bottom, ime.bottom)
            updated.marginEnd = initialEnd + systemBars.right
            updated.rightMargin = initialRight + systemBars.right
            target.layoutParams = updated
            insets
        }

        ViewCompat.requestApplyInsets(view)
    }

    private fun bindInputHandlers() {
        selectDateButton.setOnClickListener { showDatePicker() }

        towPilotInput.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                viewModel.updateTowPilotName(ContactListManager.normalizeSuggestionLabel(s?.toString().orEmpty()))
            }

            override fun afterTextChanged(s: Editable?) = Unit
        })

        towPilotInput.setOnItemClickListener { _, _, _, _ ->
            val normalized = ContactListManager.normalizeSuggestionLabel(towPilotInput.text?.toString().orEmpty())
            towPilotInput.setText(normalized, false)
            towPilotInput.setSelection(normalized.length)
            viewModel.updateTowPilotName(normalized)
        }

        towPlaneInput.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) = Unit
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                if (isUpdatingTowPlaneText) return

                val normalized = s?.toString().orEmpty().uppercase(Locale.ROOT)
                if (normalized != s?.toString().orEmpty()) {
                    isUpdatingTowPlaneText = true
                    towPlaneInput.setText(normalized)
                    towPlaneInput.setSelection(normalized.length)
                    isUpdatingTowPlaneText = false
                }
                viewModel.updateTowPlane(normalized)
            }

            override fun afterTextChanged(s: Editable?) = Unit
        })

        startNewDayButton.setOnClickListener {
            val bundle = viewModel.startNewDay()
            openDayOverview(bundle)
        }

        resumeDayButton.setOnClickListener {
            val bundle = viewModel.resumeDay()
            if (bundle.isEmpty) {
                Toast.makeText(this, getString(R.string.main_could_not_load_selected_log), Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }
            openDayOverview(bundle)
        }
    }

    private fun bindViewModel() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.selectedDate.collect { selectedDate ->
                        val formattedDate = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
                            .format(selectedDate)
                        selectDateButton.text = getString(R.string.main_selected_date_value, formattedDate)
                    }
                }

                launch {
                    viewModel.towPilotName.collect { pilotName ->
                        if (towPilotInput.text?.toString() != pilotName) {
                            towPilotInput.setText(pilotName)
                            towPilotInput.setSelection(pilotName.length)
                        }
                    }
                }

                launch {
                    viewModel.towPlane.collect { towPlane ->
                        if (towPlaneInput.text?.toString() != towPlane) {
                            towPlaneInput.setText(towPlane)
                            towPlaneInput.setSelection(towPlane.length)
                        }
                    }
                }

                launch {
                    viewModel.contacts.collect { contacts ->
                        contactAdapter.clear()
                        contactAdapter.addAll(contacts.map { ContactListManager.toSuggestionLabel(it.name, it.hasAccount) })
                        contactAdapter.notifyDataSetChanged()
                    }
                }

                launch {
                    viewModel.selectedTowPilot.collect { pilot ->
                        if (pilot == null) {
                            towPilotCheckmark.visibility = View.GONE
                        } else {
                            towPilotCheckmark.visibility = View.VISIBLE
                            towPilotCheckmark.setImageResource(
                                if (pilot.hasAccount) R.mipmap.green_checkmark else R.mipmap.new_icon_blue
                            )
                        }
                    }
                }

                launch {
                    viewModel.foundDayLog.collect { foundDayLog ->
                        resumeDayButton.visibility = if (foundDayLog) View.VISIBLE else View.GONE
                        startNewDayButton.visibility = if (foundDayLog) View.GONE else View.VISIBLE
                    }
                }
            }
        }
    }

    private fun showDatePicker() {
        val selectedDate = viewModel.selectedDate.value
        val calendar = Calendar.getInstance().apply { time = selectedDate }
        DatePickerDialog(
            this,
            { _, year, month, dayOfMonth ->
                val newDate = Calendar.getInstance().apply {
                    set(year, month, dayOfMonth)
                }.time
                viewModel.updateSelectedDate(newDate)
            },
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        ).show()
    }

    private fun openDayOverview(bundle: Bundle) {
        val intent = Intent(this, DayOverviewActivity::class.java)
        intent.putExtras(bundle)
        startActivity(intent)
    }

    private fun ensureLocationPermissionOnLaunch() {
        if (hasLocationPermission()) {
            return
        }

        val shouldShowRationale = locationPermissions.any { permission ->
            ActivityCompat.shouldShowRequestPermissionRationale(this, permission)
        }

        if (shouldShowRationale) {
            AlertDialog.Builder(this)
                .setTitle(getString(R.string.location_permission_title))
                .setMessage(getString(R.string.location_permission_rationale_message))
                .setCancelable(false)
                .setPositiveButton(getString(R.string.location_permission_allow)) { _, _ ->
                    locationPermissionLauncher.launch(locationPermissions)
                }
                .setNegativeButton(getString(R.string.location_permission_not_now), null)
                .show()
        } else {
            locationPermissionLauncher.launch(locationPermissions)
        }
    }

    private fun hasLocationPermission(): Boolean {
        val fineGranted = ContextCompat.checkSelfPermission(
            this,
            Manifest.permission.ACCESS_FINE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
        val coarseGranted = ContextCompat.checkSelfPermission(
            this,
            Manifest.permission.ACCESS_COARSE_LOCATION
        ) == PackageManager.PERMISSION_GRANTED
        return fineGranted || coarseGranted
    }

    private fun showLocationPermissionDeniedDialog() {
        AlertDialog.Builder(this)
            .setTitle(getString(R.string.location_permission_title))
            .setMessage(getString(R.string.location_permission_denied_message))
            .setPositiveButton(getString(R.string.location_permission_ok), null)
            .show()
    }

    private fun showLocationPermissionPermanentlyDeniedDialog() {
        AlertDialog.Builder(this)
            .setTitle(getString(R.string.location_permission_title))
            .setMessage(getString(R.string.location_permission_permanently_denied_message))
            .setPositiveButton(getString(R.string.location_permission_ok), null)
            .show()
    }

    private fun showPreviousLogsDialog() {
        val availableLogs = availableDayLogs()
        if (availableLogs.isEmpty()) {
            Toast.makeText(this, getString(R.string.main_no_previous_logs), Toast.LENGTH_SHORT).show()
            return
        }

        AlertDialog.Builder(this)
            .setTitle(getString(R.string.main_previous_logs_title))
            .setItems(availableLogs.map { it.displayName }.toTypedArray()) { _, which ->
                val selectedLog = availableLogs.getOrNull(which) ?: return@setItems
                val selectedDate = parseDayLogSuffix(selectedLog.fileNameSuffix)
                if (selectedDate == null) {
                    Toast.makeText(this, getString(R.string.main_unable_to_parse_log_date), Toast.LENGTH_SHORT).show()
                    return@setItems
                }

                viewModel.updateSelectedDate(selectedDate)
                val bundle = viewModel.resumeDay()
                if (bundle.isEmpty) {
                    Toast.makeText(this, getString(R.string.main_could_not_load_selected_log), Toast.LENGTH_SHORT).show()
                    return@setItems
                }

                val intent = Intent(this, DayOverviewActivity::class.java)
                intent.putExtras(bundle)
                startActivity(intent)
            }
            .show()
    }

    private fun availableDayLogs(): List<AvailableDayLog> {
        // With fixed-width yyyy_MM_dd suffixes, descending lexical order yields newest-first logs.
        val files = filesDir.list()?.toList().orEmpty().sortedDescending()
        return files
            .filter { it.startsWith(dayLogFileNamePrefix) }
            .map { fileName ->
                val suffix = fileName.removePrefix(dayLogFileNamePrefix)
                AvailableDayLog(
                    fileNameSuffix = suffix,
                    displayName = formatDayLogNameForDisplay(fileName, suffix)
                )
            }
    }

    private fun formatDayLogNameForDisplay(fileName: String, suffix: String): String {
        val parser = SimpleDateFormat("yyyy_MM_dd", Locale.ENGLISH)
        val isoDateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.ENGLISH)
        val weekdayFormatter = SimpleDateFormat("EEEE", Locale.getDefault())
        return try {
            val parsed = parser.parse(suffix)
            if (parsed == null) {
                "Parse error, name: $fileName"
            } else {
                "${isoDateFormatter.format(parsed)} (${weekdayFormatter.format(parsed)})"
            }
        } catch (_: Exception) {
            "Parse error, name: $fileName"
        }
    }

    private fun parseDayLogSuffix(fileNameSuffix: String): Date? {
        return try {
            val parser = SimpleDateFormat("yyyy_MM_dd", Locale.ENGLISH)
            parser.parse(fileNameSuffix)
        } catch (_: Exception) {
            null
        }
    }
}
